### PR TITLE
refactor!: Migrate users + agent information into storage connectors 

### PR DESCRIPTION
**Pull request by sarahwooders**
**Please describe the purpose of this pull request.**

- Refactor `Config` to `User` record
- Refactor `AgentConfig` to `AgentState` record
- Refactor `Agent` init and save to use `AgentState`
 
---

Todos 
- [x] Add `user_id` field to commands (default to `config.anon_clientid` if not provided)
  - [x] Data loading (`cli_load.py`) 
  - [x] Configuration (`cli_configure.py`) 
    - [x] Create row in `User` table if `config.anon_clientid` does not exist 
  - [x] MemGPT `run` (`cli.py`) 
    - [x] Lookup agents from metadata store, and insert new `AgentState` objects if new one is created 
  - [x] Metadata listing (e.g. listing sources, agents, etc.) 
- [x] Migrate all references to `StorageConnector` to pass in `user_id`, and optionally `agent_id` (if recall/archival memory)
- [x] Check that `run` and `configure` work
- [x] Ensure tests are all working 
- [x] Update tests to use new metadata store + storage connectors 

Remaining todos: 
* Adapt `record.text` to be `record.content`, which can be a List of strings. 
* Update all instances of `create_agent_from_preset` to use correct format
* Remove all references of `AgentConfig` and replace them with `AgentState`, which can be looked up from the metadata store 

---

**How to test**

- Regression test (unit tests should cover)
- Make sure that persistence manager is persisting (manual DB inspection and/or save+load)

**Have you tested this PR?**

WIP

### CODE REVIEW

File: memgpt/agent.py [hunk 1]
```diff
@@ -194,14 +203,9 @@ def __init__(
...
-        # - set_messages
-        # - get_messages
-        # - append_to_messages
-        self.persistence_manager = persistence_manager
-        if persistence_manager_init:
-            # creates a new agent object in the database
-            self.persistence_manager.init(self)
+        # Create the persistence manager object based on the AgentState info
+        # TODO
+        self.persistence_managager = None
```
Comment (cpacker) [inline memgpt/agent.py:208]: @sarahwooders can you add the needed code here?

File: memgpt/agent.py [hunk 2]
```diff
@@ -716,31 +616,94 @@ def heartbeat_is_paused(self):
...
-            return response
-        except Exception as e:
-            raise e
+        # Swap the system message out
+        self._swap_system_message(new_system_message)
+
+    def save(self):
+        """Save agent state locally"""
+
+        # TODO SAVE
```
Comment (cpacker) [inline memgpt/agent.py:639]: @sarahwooders can you also add the code here to do the DB write?

File: memgpt/cli/cli.py [hunk 3]
```diff
@@ -440,6 +441,9 @@ def run(
     else:  # create new agent
         # create new agent config: override defaults with args if provided
         typer.secho("\n🧬 Creating new agent...", fg=typer.colors.WHITE)
+
+        # TODO(swooders) complete hack to get the defaults loaded into AgentState properly
```
Comment (cpacker) [inline memgpt/cli/cli.py:445]: complete hack to get the defaults loaded into AgentState properly

File: memgpt/cli/cli.py [hunk 4]
```diff
@@ -451,50 +455,69 @@ def run(
...
+            name=agent_config.name,
+            persona_file=agent_config.persona,
+            human_file=agent_config.human,
+            preset=agent_config.preset,
+            model=agent_config.model,
+            model_wrapper=agent_config.model_wrapper,
+            model_endpoint_type=agent_config.model_endpoint_type,
+            model_endpoint=agent_config.model_endpoint,
+            context_window=agent_config.context_window,
+            # TODO(swooders) add embedding data properly
```
Comment (cpacker) [inline memgpt/cli/cli.py:471]: TODO add embedding data properly

File: memgpt/cli/cli.py [hunk 5]
```diff
@@ -451,50 +455,69 @@ def run(
...
+        )
+
+        typer.secho(f"->  🤖 Using persona profile '{agent_init_state.persona_file}'", fg=typer.colors.WHITE)
+        typer.secho(f"->  🧑 Using human profile '{agent_init_state.human_file}'", fg=typer.colors.WHITE)
 
         # Supress llama-index noise
-        with suppress_stdout():
-            # TODO: allow configrable state manager (only local is supported right now)
-            persistence_manager = LocalStateManager(agent_config)  # TODO: insert dataset/pre-fill
+        # TODO(swooders) add persistence manager code? or comment out?
```
Comment (cpacker) [inline memgpt/cli/cli.py:486]: TODO add persistence manager code? or comment out?

File: memgpt/cli/cli.py [hunk 6]
```diff
@@ -451,50 +455,69 @@ def run(
...
-                utils.get_persona_text(agent_config.persona),
-                utils.get_human_text(agent_config.human),
-                interface,
-                persistence_manager,
+            memgpt_agent = presets.create_agent_from_preset(
+                agent_state=agent_init_state,
+                interface=interface,
             )
         except ValueError as e:
+            # TODO(swooders) what's the equivalent cleanup code for the new DB refactor?
```
Comment (cpacker) [inline memgpt/cli/cli.py:498]: TODO what's the equivalent cleanup code for the new DB refactor?

File: memgpt/data_types.py [hunk 7]
```diff
@@ -146,3 +147,146 @@ def __init__(
...
+        embedding_model: Optional[str] = None,
+        embedding_dim: Optional[int] = None,  # TODO(swooders) similarly these probably should not be optional
+        embedding_chunk_size: Optional[int] = None,
+        # other
+        preset: Optional[str] = None,
+        data_sources: Optional[list] = None,
+        create_time: Optional[str] = None,
+        memgpt_version: Optional[str] = None,
+    ):
+        # TODO(swooders) we need to handle the case where name is None here
```
Comment (cpacker) [inline memgpt/data_types.py:260]: TODO how should we handle defaults here? exact same as agentconfig?

File: memgpt/data_types.py [hunk 8]
```diff
@@ -146,3 +147,146 @@ def __init__(
...
+        self.embedding_endpoint = embedding_endpoint
+        self.embedding_model = embedding_model
+        self.embedding_dim = embedding_dim
+        self.embedding_chunk_size = embedding_chunk_size
+
+        # other
+        # NOTE: preset is only used to determine an initial combination of system message + functions (can be None)
+        self.preset = (
+            DEFAULT_PRESET if preset is None else preset
+        )  # TODO(swooders) we should probably allow this to be None? what if someone wants to create w/o a preset?
```
Comment (cpacker) [inline memgpt/data_types.py:286]: TODO allow preset is None?

File: memgpt/data_types.py (hunk 7 above)
Comment (sarahwooders) [reply memgpt/data_types.py:260]: What do you think of not allowing any defaults, to enforce that everything is passed in from the `User` record (which defines all the default values)?

File: memgpt/config.py [hunk 9]
```diff
@@ -31,6 +31,120 @@ def set_field(config, section, field, value):
     config.set(section, field, value)
 
 
+@dataclass
+class Config:
```
Comment (sarahwooders) [inline memgpt/config.py:35]: Ignore this, decided to just keep `MemGPTConfig` for now and maybe gradually migrate to this

File: memgpt/agent_store/local.py [hunk 10]
```diff
@@ -252,9 +252,11 @@ def json_to_message(self, message_json) -> Message:
             user_id=self.config.anon_clientid,
             agent_id=self.agent_config.name,
             role=message["role"],
+            name=message["name"] if "name" in message else None,
             text=message["content"],
             model=self.agent_config.model,
             created_at=parse_formatted_time(timestamp),
+            # TODO(swooders) this should be tools, right?
```
Comment (cpacker) [inline memgpt/agent_store/local.py:259]: @sarahwooders these should be revised to the tool-call schema, right?

File: memgpt/persistence_manager.py [hunk 11]
```diff
@@ -133,11 +131,11 @@ def json_to_message(self, message_json) -> Message:
             tool_calls = None
 
         return Message(
-            user_id=self.config.anon_clientid,
-            agent_id=self.agent_config.name,
+            user_id=self.agent_state.user_id,
+            agent_id=self.agent_state.id,
             role=message["role"],
             text=message["content"],
```
Comment (cpacker) [inline memgpt/persistence_manager.py:137]: @sarahwooders Also, I think we should be consistent with OAI and make the naming `text` -> `content` in the data type.

Relatedly, in the OAI API spec, UserMessage is slightly different than a regular Message, in that `Message.content` can be either a string (as per usual), or an array. This is to support multi-modality:
image

This is something we might want to consider baking into the schema on the initial refactor release (0.2.12).

File: memgpt/cli/cli.py (hunk 3 above)
Comment (sarahwooders) [reply memgpt/cli/cli.py]: I think we can remove `AgentConfig` here?

File: memgpt/agent_store/local.py (hunk 10 above)
Comment (sarahwooders) [reply memgpt/agent_store/local.py]: removed this file

**Review by cpacker — APPROVED** [review]

File: memgpt/cli/cli.py (hunk 3 above)
Comment (cpacker) [reply memgpt/cli/cli.py]: Punted to later PR:
> Remove all references of AgentConfig and replace them with AgentState, which can be looked up from the metadata store

**Pull request merged**